jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -1601,6 +1601,190 @@ ecma_builtin_typedarray_prototype_find_index (ecma_value_t this_arg, /**< this a
   return ecma_builtin_typedarray_prototype_find_helper (this_arg, predicate, predicate_this_arg, false);
 } /* ecma_builtin_typedarray_prototype_find_index */
 
+/**
+ * The %TypedArray%.prototype object's 'copyWithin' routine helper
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static uint32_t
+ecma_builtin_typedarray_prototype_copy_within_helper(uint32_t length, /**< length of the typedarray */
+                                                     ecma_number_t index, /**< copyWithin target or start or end property */
+                                                     bool is_end_index ) /**< true - normalize the end index */
+{
+  uint32_t ret_value;
+
+  if (!ecma_number_is_nan (index))
+  {
+    if (ecma_number_is_zero (index))
+    {
+      ret_value = 0;
+    }
+    else if (ecma_number_is_infinity (index))
+    {
+      ret_value = ecma_number_is_negative (index) ? 0 : length;
+    }
+    else
+    {
+      if (ecma_number_is_negative (index))
+      {
+        ecma_number_t index_neg = -index;
+        int32_t norm_index = (int32_t) ((int32_t)length - (int32_t) ecma_number_to_int32 (index_neg));
+        ret_value = norm_index > 0 ? (uint32_t) norm_index : 0;
+      }
+      else
+      {
+        ret_value = ecma_number_to_uint32 (index) > length ? length : ecma_number_to_uint32 (index);
+      }
+    }
+  }
+  else
+  {
+    ret_value = is_end_index ? length : 0;
+  }
+
+  return ret_value;
+} /* ecma_builtin_typedarray_prototype_copy_within_helper */
+
+/**
+ * The %TypedArray%.prototype object's 'copyWithin' routine
+ *
+ * See also:
+ *          ECMA-262 v6, 22.2.3.5
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_typedarray_prototype_copy_within(ecma_value_t this_arg, /**< this argument */
+                                              const ecma_value_t args[], /**< arguments list */
+                                              ecma_length_t args_number) /**< number of arguments */
+{
+  if (!ecma_is_typedarray (this_arg))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a TypedArray."));
+  }
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (this_arg);
+  uint32_t length = ecma_typedarray_get_length (typedarray_p);
+  uint32_t target = 0;
+  uint32_t start = 0;
+  uint32_t end = length;
+  ecma_number_t num_var;
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+
+  if (args_number > 0)
```

Comment:
The `arg[0]`, `arg[1]` and `arg[2]` pre-processing looks very similar, would it be possible to have utility method to reduce the code duplication?

Reply:
In addition, there should be tests for the error cases for the arguments